lib/stages/VisWriter.cpp

```diff
@@ -0,0 +1,379 @@
+    // Check the dataset ID hasn't changed
+    if (acqs.count(dataset_id) == 0) {
+        init_acq(dataset_id, make_metadata(dataset_id));
+    }
+
+    // Get the acquisition we are writing into
+    auto& acq = *(acqs.at(dataset_id));
+    acqs_lock.unlock();
+
+    // If the dataset is bad, skip the frame and move onto the next
+    if (acq.bad_dataset) {
+        bad_dataset_frame_counter.labels({dataset_id.to_string()}).inc();
+
+        // Check if the frequency we are receiving is on the list of frequencies
+        // we are processing
+        // TODO: this should probably be reported to prometheus
+    } else if (acq.freq_id_map.count(freq_id) == 0) {
+        WARN("Frequency id={:d} not enabled for VisWriter, discarding frame", freq_id);
+
+        // Check that the number of visibilities matches what we expect
+    } else if (frame.num_prod != acq.num_vis) {
+        FATAL_ERROR("Number of products in frame doesn't match state or file ({:d} != {:d}).",
+                    frame.num_prod, acq.num_vis);
+        return;
+
+    } else {
+
+        // Get the time and frequency of the frame
+        time_ctype t = {std::get<0>(ftime), ts_to_double(std::get<1>(ftime))};
+        uint32_t freq_ind = acq.freq_id_map.at(freq_id);
+
+        // Add all the new information to the file.
+        bool late;
+        double start = current_time();
+
+        // Lock and write data
+        {
+            std::lock_guard<std::mutex> lock(write_mutex);
+            late = acq.file_bundle->add_sample(t, freq_ind, frame);
+        }
+        acq.last_update = current_time();
+        double elapsed = acq.last_update - start;
+
+        DEBUG("Written frequency {:d} in {:.5f} s", freq_id, elapsed);
+
+        // Increase metric count if we dropped a frame at write time
+        if (late) {
+            late_frame_counter.labels({std::to_string(freq_id)}).inc();
+        }
+
+        // Update average write time in prometheus
+        write_time.add_sample(elapsed);
+        write_time_metric.set(write_time.average());
+    }
+}
+
+void VisWriter::init_acq(dset_id_t ds_id, std::map<std::string, std::string> metadata) {
+
+    // Create the new acqState
+    auto fp = datasetManager::instance().fingerprint(ds_id, critical_state_types);
+
+    // If the fingerprint is already known, we don't need to start a new
+    // acquisition, just add a map from the dataset_id to the acquisition we
+    // should use.
+    if (acqs_fingerprint.count(fp) > 0) {
+        INFO("Got new dataset ID={} with known fingerprint={}.", ds_id, fp);
+        acqs[ds_id] = acqs_fingerprint.at(fp);
+        return;
+    }
+
+    INFO("Got new dataset ID={} with new fingerprint={}. Creating new acquisition.", ds_id, fp);
+
+    // If it is not known we need to initialise everything
+    acqs_fingerprint[fp] = std::make_shared<acqState>();
+    acqs[ds_id] = acqs_fingerprint.at(fp);
+    auto& acq = *(acqs.at(ds_id));
+
+    // get dataset states
+    get_dataset_state(ds_id);
```

Comment:
@jrs65 Does `get_dataset_state()` need to be called here? Because if not, I can move `init_acq` into `BaseWriter`.

Reply:
I think so, it's needed to initialise information about the acquisition (which it sets within `acqs[ds_id]`). Is there a good other place to call it from?

Why can't you add `get_dataset_state` as a virtual method on `BaseWriter` and have it no-op for HFBWriter?

Reply:
I've already added `get_dataset_state` as a virtual method to `BaseWriter`and added code to the method in `HFBWriter`